File: rust/src/api/tags.rs
use std::path::Path;

use anyhow::{anyhow, Result};

use crate::db;

fn key_from_bytes(bytes: Vec<u8>) -> Result<[u8; 32]> {
    if bytes.len() != 32 {
        return Err(anyhow!("invalid key length"));
    }
    let mut key = [0u8; 32];
    key.copy_from_slice(&bytes);
    Ok(key)
}

#[flutter_rust_bridge::frb]
pub fn db_list_tags(app_dir: String, key: Vec<u8>) -> Result<Vec<db::Tag>> {
    let key = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::list_tags(&conn, &key)
}

#[flutter_rust_bridge::frb]
pub fn db_upsert_tag(app_dir: String, key: Vec<u8>, name: String) -> Result<db::Tag> {
    let key = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::upsert_tag(&conn, &key, &name)
}

#[flutter_rust_bridge::frb]
pub fn db_list_message_tags(
    app_dir: String,
    key: Vec<u8>,
    message_id: String,
) -> Result<Vec<db::Tag>> {
    let key = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::list_message_tags(&conn, &key, &message_id)
}

#[flutter_rust_bridge::frb]
pub fn db_set_message_tags(
    app_dir: String,
    key: Vec<u8>,
    message_id: String,
    tag_ids: Vec<String>,
) -> Result<Vec<db::Tag>> {
    let key = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::set_message_tags(&conn, &key, &message_id, &tag_ids)
}

#[flutter_rust_bridge::frb]
pub fn db_list_message_suggested_tags(
    app_dir: String,
    key: Vec<u8>,
    message_id: String,
) -> Result<Vec<String>> {
    let key = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::list_message_suggested_tags(&conn, &key, &message_id)
}

#[flutter_rust_bridge::frb]
pub fn db_list_tag_merge_suggestions(
    app_dir: String,
    key: Vec<u8>,
    limit: u32,
) -> Result<Vec<db::TagMergeSuggestion>> {
    let key = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::list_tag_merge_suggestions(&conn, &key, limit as usize)
}

#[flutter_rust_bridge::frb]
pub fn db_merge_tags(
    app_dir: String,
    key: Vec<u8>,
    source_tag_id: String,
    target_tag_id: String,
) -> Result<u32> {
    let key = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::merge_tags(&conn, &key, &source_tag_id, &target_tag_id)
}

#[flutter_rust_bridge::frb]
pub fn db_record_tag_merge_feedback(
    app_dir: String,
    key: Vec<u8>,
    source_tag_id: String,
    target_tag_id: String,
    reason: String,
    action: String,
) -> Result<()> {
    let _ = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::record_tag_merge_feedback(&conn, &source_tag_id, &target_tag_id, &reason, &action)
}

#[flutter_rust_bridge::frb]
pub fn db_list_message_ids_by_tag_ids(
    app_dir: String,
    key: Vec<u8>,
    conversation_id: String,
    tag_ids: Vec<String>,
) -> Result<Vec<String>> {
    let _ = key_from_bytes(key)?;
    let conn = db::open(Path::new(&app_dir))?;
    db::list_message_ids_by_tag_ids(&conn, &conversation_id, &tag_ids)
}

File: lib/core/runtime_compat/api/tags.dart
import '../../models/app_models.dart';

Future<List<Tag>> dbListTags(
        {required String appDir, required List<int> key}) =>
    throw UnsupportedError('rust_runtime_removed:dbListTags');

Future<Tag> dbUpsertTag(
        {required String appDir,
        required List<int> key,
        required String name}) =>
    throw UnsupportedError('rust_runtime_removed:dbUpsertTag');

Future<List<Tag>> dbListMessageTags(
        {required String appDir,
        required List<int> key,
        required String messageId}) =>
    throw UnsupportedError('rust_runtime_removed:dbListMessageTags');

Future<List<Tag>> dbSetMessageTags(
        {required String appDir,
        required List<int> key,
        required String messageId,
        required List<String> tagIds}) =>
    throw UnsupportedError('rust_runtime_removed:dbSetMessageTags');

Future<List<String>> dbListMessageSuggestedTags(
        {required String appDir,
        required List<int> key,
        required String messageId}) =>
    throw UnsupportedError('rust_runtime_removed:dbListMessageSuggestedTags');

Future<List<String>> dbListManualMessageTagNames(
        {required String appDir,
        required List<int> key,
        required String messageId}) =>
    throw UnsupportedError('rust_runtime_removed:dbListManualMessageTagNames');

Future<List<TagMergeSuggestion>> dbListTagMergeSuggestions(
        {required String appDir, required List<int> key, required int limit}) =>
    throw UnsupportedError('rust_runtime_removed:dbListTagMergeSuggestions');

Future<int> dbMergeTags(
        {required String appDir,
        required List<int> key,
        required String sourceTagId,
        required String targetTagId}) =>
    throw UnsupportedError('rust_runtime_removed:dbMergeTags');

Future<List<TagMergeSuggestion>> dbListHiddenTagMergeSuggestions(
        {required String appDir, required List<int> key, required int limit}) =>
    throw UnsupportedError(
        'rust_runtime_removed:dbListHiddenTagMergeSuggestions');

Future<void> dbDeleteTag(
        {required String appDir,
        required List<int> key,
        required String tagId}) =>
    throw UnsupportedError('rust_runtime_removed:dbDeleteTag');

Future<void> dbRecordTagMergeFeedback(
        {required String appDir,
        required List<int> key,
        required String sourceTagId,
        required String targetTagId,
        required String reason,
        required String action}) =>
    throw UnsupportedError('rust_runtime_removed:dbRecordTagMergeFeedback');

Future<void> dbClearTagMergeFeedback(
        {required String appDir,
        required List<int> key,
        required String sourceTagId,
        required String targetTagId}) =>
    throw UnsupportedError('rust_runtime_removed:dbClearTagMergeFeedback');

Future<List<String>> dbListMessageIdsByTagIds(
        {required String appDir,
        required List<int> key,
        required String conversationId,
        required List<String> tagIds}) =>
    throw UnsupportedError('rust_runtime_removed:dbListMessageIdsByTagIds');
